system/iptables/iptables.c

```diff
@@ -0,0 +1,462 @@
+ * apps/system/iptables/iptables.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include <nuttx/net/netfilter/ip_tables.h>
+#include <nuttx/net/netfilter/nf_nat.h>
+
+#include "argtable3.h"
+#include "netutils/netlib.h"
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/* TODO: Our getopt(), which argtable3 depends, does not support nonoptions
+ * in the middle, so commands like 'iptables -I chain rulenum -j MASQUERADE'
+ * cannot be supported, because the 'rulenum' will stop getopt() logic and
+ * the -j option will never be parsed. Although we write 'rulenum' option
+ * here, it may not work well, but it will work when our getopt() is updated.
+ */
+
+struct iptables_args_s
+{
+  FAR struct arg_str *table;
+
+  FAR struct arg_str *append_chain;
+  FAR struct arg_str *insert_chain;
+  FAR struct arg_str *delete_chain;
+  FAR struct arg_str *flush_chain;
+  FAR struct arg_str *list_chain;
+
+  FAR struct arg_int *rulenum;
+
+  FAR struct arg_str *target;
+  FAR struct arg_str *outifname;
+
+  FAR struct arg_end *end;
+};
+
+enum iptables_command_e
+{
+  COMMAND_APPEND,
+  COMMAND_INSERT,
+  COMMAND_DELETE,
+  COMMAND_FLUSH,
+  COMMAND_LIST,
+  COMMAND_NUM,
+};
+
+struct iptables_command_s
+{
+  enum iptables_command_e cmd;
+  enum nf_inet_hooks hook;
+  int rulenum;
+};
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static FAR const char *g_hooknames[] =
+{
+  [NF_INET_PRE_ROUTING]  = "PREROUTING",
+  [NF_INET_LOCAL_IN]     = "INPUT",
+  [NF_INET_FORWARD]      = "FORWARD",
+  [NF_INET_LOCAL_OUT]    = "OUTPUT",
+  [NF_INET_POST_ROUTING] = "POSTROUTING",
```

Comment:
can we have c89 compatible init?

Reply:
Also changed into one line since we don't use macro to control them.